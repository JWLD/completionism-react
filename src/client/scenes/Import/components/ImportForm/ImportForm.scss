@import 'global';

$inputHeight: 4rem;
$boxPadding: 2rem;
$wrapWidth: 70rem;

.import-page {
	.sctn:not(:last-child) {
		margin-bottom: $pagePadding * .5;
	}

	.sctn__header {
		justify-content: flex-start;
		margin-bottom: $boxPadding * .25;
		color: white;
		font: 4rem DinRC;
	}

	.sctn__input-wrap {
		align-items: stretch;
		width: $wrapWidth;
		padding: $boxPadding;
		border-radius: .4rem;
		background-color: #333;
	}
}
