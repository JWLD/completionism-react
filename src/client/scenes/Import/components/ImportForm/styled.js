import styled from 'styled-components';

export const FormSection = styled.section`
	:not(:last-child) {
		margin-bottom: 3rem;
	}
`;

export const FormHeader = styled.h2`
	margin-bottom: 1rem;
	color: white;
	font: 4rem DinRC;
`;

export const InputWrap = styled.div`
	align-items: stretch;
	width: 70rem;
	padding: 2rem;
	border-radius: .4rem;
	background-color: #333;
`;
