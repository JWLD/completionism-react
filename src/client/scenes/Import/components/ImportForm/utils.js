import { fetchBattleNetCharacterData } from 'services/api'

const extractCategoriesArray = categories => {
  return Object.keys(categories).reduce((categoryNamesArray, key) => {
    const categoryIsSetToTrue = categories[key]

    if (categoryIsSetToTrue) {
      categoryNamesArray.push(key)
    }

    return categoryNamesArray
  }, [])
}

const constructUrlParams = ({ categories, character }) => ({
  cats: extractCategoriesArray(categories),
  name: character.name,
  realm: character.realm,
  region: character.region
})

const transformPetDataForStorage = petData => {
  return petData.reduce((acc, pet) => {
    acc[pet.id] = {
      level: pet.level,
      quality: pet.quality
    }

    return acc
  }, {})
}

const saveCharDataToLocalStorage = ({ char, collections }) => {
  return Object.keys(collections).map(key => {
    const collection =
      key === 'pets'
        ? transformPetDataForStorage(collections[key])
        : collections[key]

    localStorage[key] = JSON.stringify({ char, collection })
  })
}

export const fetchCharData = async values => {
  const urlParams = constructUrlParams(values)
  const characterData = await fetchBattleNetCharacterData(urlParams)

  return saveCharDataToLocalStorage(characterData)
}
