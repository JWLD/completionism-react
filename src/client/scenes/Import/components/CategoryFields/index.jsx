import React from 'react'
import { Field } from 'redux-form'

import { CATEGORIES_BY_SECTION as CATEGORIES } from 'constants/categories'

import * as SC from 'CategoryFields/styled'
import { CheckboxField } from 'formFields'

const CategoryFields = () => {
  const renderInputRows = categories => {
    return categories.map(({ battleNet, key, name }) => {
      if (!battleNet) return null

      return (
        <SC.CheckboxLabel key={key}>
          {name}
          <Field
            component={CheckboxField}
            name={key}
            StyledComponent={SC.CheckboxInput}
          />
          <SC.FakeInput className="fa fa-check" />
        </SC.CheckboxLabel>
      )
    })
  }

  const renderCheckboxColumns = () => {
    return Object.keys(CATEGORIES).map(category => (
      <SC.CheckboxWrap key={category}>
        {renderInputRows(CATEGORIES[category])}
      </SC.CheckboxWrap>
    ))
  }

  return renderCheckboxColumns()
}

export default CategoryFields
