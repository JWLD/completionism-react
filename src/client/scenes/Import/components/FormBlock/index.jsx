import React from 'react'
import PropTypes from 'prop-types'
import { FormSection } from 'redux-form'

import * as SC from 'FormBlock/styled'

const FormBlock = props => (
  <SC.FormSection>
    <SC.FormHeader>{props.header}</SC.FormHeader>

    <SC.InputWrap>
      <FormSection name={props.name}>{props.fieldsComponent}</FormSection>
    </SC.InputWrap>
  </SC.FormSection>
)

FormBlock.propTypes = {
  fieldsComponent: PropTypes.object.isRequired,
  header: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired
}

export default FormBlock
