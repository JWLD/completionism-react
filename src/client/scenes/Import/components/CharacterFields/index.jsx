import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Field } from 'redux-form'

import { realmDataSelector, regionSelector } from './selectors'
import { REGIONS } from 'constants/blizzard'
import { fetchRealmData } from 'scenes/Import/actions'

import * as SC from './styled'
import { InputField, SelectBoxField } from 'components/ReduxFields'

class CharacterFields extends Component {
  componentDidMount() {
    this.getRealmData(this.props.region)
  }

  onRegionChange = (changeEvent) => {
    const newRegion = changeEvent.target.value

    this.getRealmData(newRegion)
  }

  getRealmData(region) {
    if (!this.props.realmList[region]) {
      this.props.fetchRealmData(region)
    }
  }

  render() {
    return (
      <SC.CharacterFieldsWrap>
        <Field
          component={SelectBoxField}
          name="region"
          onChange={this.onRegionChange}
          options={REGIONS}
          StyledComponent={SC.RegionSelect}
        />

        <Field
          component={SelectBoxField}
          name="realm"
          options={this.props.realmData}
          placeholder="Select Realm"
          StyledComponent={SC.RealmSelect}
        />

        <Field component={InputField} name="name" placeholder="Name" StyledComponent={SC.NameInput} />
      </SC.CharacterFieldsWrap>
    )
  }
}

CharacterFields.propTypes = {
  fetchRealmData: PropTypes.func.isRequired,
  realmData: PropTypes.object.isRequired,
  realmList: PropTypes.object.isRequired,
  region: PropTypes.string.isRequired
}

const mapStateToProps = state => ({
  realmData: realmDataSelector(state),
  realmList: state.import.realms,
  region: regionSelector(state)
})

const mapDispatchToProps = {
  fetchRealmData
}

export default connect(mapStateToProps, mapDispatchToProps)(CharacterFields)
