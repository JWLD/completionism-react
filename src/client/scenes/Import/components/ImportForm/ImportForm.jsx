import React from 'react';
import { reduxForm } from 'redux-form';

import CategoryFields from './components/CategoryFields/CategoryFields';
import CharacterFields from './components/CharacterFields/CharacterFields';
import { FormHeader, FormSection, InputWrap } from './styled';

const ImportForm = () => (
	<form className="import-form">
		<FormSection>
			<FormHeader>Select Categories</FormHeader>

			<InputWrap className="sctn__input-wrap">
				<CategoryFields />
			</InputWrap>
		</FormSection>

		<FormSection>
			<FormHeader>Select Character</FormHeader>

			<InputWrap className="sctn__input-wrap">
				<CharacterFields />
			</InputWrap>
		</FormSection>
	</form>
);

export default reduxForm({ form: 'import' })(ImportForm);
