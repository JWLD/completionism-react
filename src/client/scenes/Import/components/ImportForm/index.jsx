import React from 'react'
import PropTypes from 'prop-types'
import { reduxForm } from 'redux-form'

import { fetchCharData } from './utils'

import FormBlock from 'FormBlock'
import CategoryFields from 'CategoryFields'
import CharacterFields from 'CharacterFields'

export const ImportForm = props => (
  <form onSubmit={props.handleSubmit(fetchCharData)}>
    <FormBlock
      header="Select Categories"
      name="categories"
      fieldsComponent={<CategoryFields />}
    />
    <FormBlock
      header="Select Character"
      name="character"
      fieldsComponent={<CharacterFields />}
    />
  </form>
)

ImportForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
}

const FormWrapper = reduxForm({
  form: 'import',
  initialValues: {
    character: {
      region: 'eu'
    }
  }
})

export default FormWrapper(ImportForm)
