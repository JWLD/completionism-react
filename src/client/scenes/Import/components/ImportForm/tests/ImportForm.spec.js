import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { ImportForm } from 'scenes/Import/components/ImportForm'

Enzyme.configure({ adapter: new Adapter() })

const mockProps = {
  handleSubmit: jest.fn()
}

describe('<ImportForm />', () => {
  it('renders correctly', () => {
    const component = shallow(<ImportForm {...mockProps} />)

    expect(component).toMatchSnapshot()
  })
})
