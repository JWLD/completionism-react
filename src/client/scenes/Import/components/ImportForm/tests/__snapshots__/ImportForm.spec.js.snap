// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImportForm /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImportForm
    handleSubmit={
      [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <FormBlock
          fieldsComponent={<CategoryFields />}
          header="Select Categories"
          name="categories"
        />,
        <FormBlock
          fieldsComponent={<Connect(CharacterFields) />}
          header="Select Character"
          name="character"
        />,
      ],
      "onSubmit": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fieldsComponent": <CategoryFields />,
          "header": "Select Categories",
          "name": "categories",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fieldsComponent": <Connect(CharacterFields) />,
          "header": "Select Character",
          "name": "character",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <FormBlock
            fieldsComponent={<CategoryFields />}
            header="Select Categories"
            name="categories"
          />,
          <FormBlock
            fieldsComponent={<Connect(CharacterFields) />}
            header="Select Character"
            name="character"
          />,
        ],
        "onSubmit": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fieldsComponent": <CategoryFields />,
            "header": "Select Categories",
            "name": "categories",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fieldsComponent": <Connect(CharacterFields) />,
            "header": "Select Character",
            "name": "character",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
